src/binding/python/Attributable.cpp

```diff
@@ -0,0 +1,173 @@
+/* Copyright 2018 Axel Huebl
+ *
+ * This file is part of openPMD-api.
+ *
+ * openPMD-api is free software: you can redistribute it and/or modify
+ * it under the terms of of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * openPMD-api is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with openPMD-api.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <pybind11/pybind11.h>
+#include <pybind11/stl_bind.h>
+#include <pybind11/stl.h>
+
+#include "openPMD/backend/Attributable.hpp"
+#include "openPMD/backend/Attribute.hpp"
+#include "openPMD/auxiliary/Variant.hpp"
+
+#include <string>
+#include <vector>
+#include <array>
+
+
+// std::variant
+//   https://pybind11.readthedocs.io/en/stable/advanced/cast/stl.html
+namespace pybind11 {
+namespace detail {
+    template< typename... Ts >
+    struct type_caster< variantSrc::variant< Ts... > > :
+        variant_caster< variantSrc::variant< Ts... > >
+    {};
+
+} // namespace detail
+} // namespace pybind11
+
+namespace py = pybind11;
+using namespace openPMD;
+
+using PyAttributeKeys = std::vector< std::string >;
+//PYBIND11_MAKE_OPAQUE(PyAttributeKeys)
+
+void init_Attributable(py::module &m) {
+    py::class_<Attributable>(m, "Attributable")
+        .def(py::init<>())
+        .def(py::init<Attributable const &>())
+
+        .def("__repr__",
+            [](Attributable const & attr) {
+                return "<openPMD.Attributable with'" + std::to_string(attr.numAttributes()) + "' attributes>";
+            }
+        )
+
+        .def_property_readonly(
+            "attributes",
+            []( Attributable & attr )
+            {
+                return attr.attributes();
+            },
+            // ref + keepalive
+            py::return_value_policy::reference_internal
+        )
+
+        // C++ pass-through API
+        //   note: since the value is expected as T&& we have to write a lambda here
+        .def("set_attribute", []( Attributable & attr, std::string const& key, char const& v ) {
+            attr.setAttribute(key, v);
```

Comment:
@C0nsultant do you think the `T&&` signature is necessary for `setAttribute`? Or would a `T&` do as well?

Reply:
In restrospect ```T&&``` was indeed premature optimization. It improves performance when the user passes literals to the setters, but otherwise is not really useful.

I would not mourn after ```T&&``` if we make it ```T const&```.

Reply:
Ok, that allows me to write a on-liner per specialization here. Cool.